benchmarks/ddp/benchmark.py

```diff
@@ -0,0 +1,276 @@
+
+    model = benchmark.create_model()
+    data = [(benchmark.generate_inputs(), benchmark.generate_target())]
+    criterion = nn.CrossEntropyLoss()
+    optimizer = optim.SGD(model.parameters(),
+                          0.001,
+                          momentum=0.9,
+                          weight_decay=1e-4)
+    if pg.size() > 1:
+        model = torch.nn.parallel.DistributedDataParallel(
+            model,
+            device_ids=[torch.cuda.current_device()],
+            broadcast_buffers=False,
+            process_group=pg)
+
+    measurements = []
+    warmup_iterations = 5
+    measured_iterations = 10
+    for (inputs, target) in (data * (warmup_iterations + measured_iterations)):
+        start = time.time()
+        output = model(*inputs)
+        loss = criterion(output, target)
+        loss.backward()
+        optimizer.step()
+        torch.cuda.synchronize()
+        measurements.append(time.time() - start)
+
+    # Throw away measurements for warmup iterations
+    return measurements[warmup_iterations:]
+
+
+def run_benchmark(prefix, ranks, benchmark):
+    def local_print(msg):
+        if dist.get_rank() == 0:
+            print(msg, end='', flush=True)
+
+    def print_measurements(prefix, nelem, measurements):
+        measurements = sorted(measurements)
+        local_print("%s: " % prefix)
+        for p in [50, 75, 90, 95]:
+            v = np.percentile(measurements, p)
+            local_print("p%02d:  %1.3f  %4d/s  " % (p, v, nelem / v))
+        local_print("\n")
+
+    group_nccl = dist.new_group(ranks=ranks, backend="nccl")
+    group_gloo = dist.new_group(ranks=ranks, backend="gloo")
+    group_to_benchmark = group_nccl
+
+    measurements = []
+    if dist.get_rank() in set(ranks):
+        measurements = benchmark_process_group(group_to_benchmark, benchmark)
+
+    dist.destroy_process_group(group_nccl)
+    dist.destroy_process_group(group_gloo)
+    dist.barrier()
+
+    # Force destruction of everything that is no longer used.
+    # Most notably, this is needed to destroy the NCCL process group
+    # if the PyTorch 1.0 DistributedDataParallel class with Python
+    # side reduction is used. If we don't explicitly run GC,
+    # creation of NCCL process groups later on results in a crash.
+    gc.collect()
+
+    # Aggregate measurements for better estimation of percentiles
+    measurements = list(itertools.chain(*allgather_object(measurements)))
+    print_measurements(prefix, benchmark.batch_size, measurements)
+    return measurements
+
+
+def sweep(benchmark):
+    # Synthesize the set of benchmarks to run.
+    # This list contain tuples for ("string prefix", [rank...]).
+    benchmarks = []
+
+    def append_benchmark(prefix, ranks):
+        prefix = "%4d GPUs -- %s" % (len(ranks), prefix)
+        benchmarks.append((prefix, ranks))
+
+    # Every process runs once by themselves to warm up (CUDA init, etc).
+    append_benchmark("  warmup", [dist.get_rank()])
```

Comment:
Do we need this warmup benchmark? Don't we already have warmup iterations in `benchmark_process_group`?

Reply:
This full warmup iteration is to make sure every rank has performed the same warmup, before starting the real ranks. Otherwise, we'd progressively need to warm up more ranks.